### ISSUE TITLE
Dispose changes

### ISSUE THREAD

**Issue by kealist** [unosquare/ffmediaelement#371]
# Dispose changes

I just upgraded and noticed the lack of `Dispose()` on MediaElement.   How should I handle this now?  I am trying `Close()` but it is not returning and I get memory leak errors

```
2019-04-10 08:01:53,290 ERROR VideoControl.xaml.cs: 225 - Ch1 (Channel 1): Unmanaged references were left alive. This is an indication that there is a memory leak.
    CodecContext - MediaComponent.cs; Line: 79, Member: .ctor - Instances: 5
    Packet - MediaPacket.cs; Line: 83, Member: CreateReadPacket - Instances: 125
    SwsContext - VideoComponent.cs; Line: 228, Member: MaterializeFrame - Instances: 5
```

## Issue Categories

 - [ ] Bug
 - [ ] Feature Request
 - [x] Question
 - [ ] Not sure

## Version Information

 - [x] NuGet Package 2.90

**Comment by mariodivece** [comment]
WPF controls should not be disposable. That is what WPF users are expecting and that's why the Dispose method has been removed. The real issue is to find out why the Close method is not being completed successfully. Could you please give me a hand and provide a zip file with a clean project reproducing the issue?

Additionally, could you please test by building the latest master branch and see if the issue is still the same?

Thanks!

**Comment by kealist** [comment]
I can try to remove all the higher level disposes now that it isn't required, in order to see if that helps resolve the issue.    I can try in the next week to get a minimal example.
 
I'm not able to build master currently because of `warning as errors` that block the build.   I tried disabling it, I have previously had stylecop on and it seems to be the issue.    I'll look at it again but I'm under a crunch this week.

...Feature request (automated prerelease builds off master?)

**Comment by paurail** [comment]
Same here. after every `mediaPlayer.Source = new Uri(videoFile.FullName)` memory usage slightly increases and never goes back. Close doesn't held and Dispose used to..

**Comment by mariodivece** [comment]
> I can try to remove all the higher level disposes now that it isn't required, in order to see if that helps resolve the issue. I can try in the next week to get a minimal example.
> 
> I'm not able to build master currently because of `warning as errors` that block the build. I tried disabling it, I have previously had stylecop on and it seems to be the issue. I'll look at it again but I'm under a crunch this week.
> 
> ...Feature request (automated prerelease builds off master?)

Any updates here? I am not getting the effects of memory increases or memory leak errors in the output. Memory returns back to normal after closing -- sometimes to even lower.

![image](https://user-images.githubusercontent.com/351978/58763428-45529b00-8520-11e9-89d6-0b7ec5ea7a8f.png)

**Comment by AaronAnz** [comment]
mariodivece are you using an HLS stream? I am passing an m3u8 to the controls source property on view XAML. I'm experiencing a memory leak like they describe via the latest nuget package (4.1.300) as of this comment. I am not manually pre-loading ffmpeg at all. For sake of a test I made my app open my "MediaView" with a random twitch stream 10 times and this is what is observed. I'm using Ninject and MahApps in case those may interfere? Doubt it but figured i'd mention it.

![FFME_LEAK](https://user-images.githubusercontent.com/27394920/60068382-5b88ec80-96c3-11e9-9c87-d55baeb4e3b1.PNG)

```
   <Grid Grid.Row="0">
        <ffme:MediaElement x:Name="MediaPlayer" LoadedBehavior="Play" UnloadedBehavior="Manual" 
                            Source="{Binding Path=PlaylistSource.Source, Converter={StaticResource StringToUriConverter}}"/>       
    </Grid>
```
In order to even get my MediaView to dispose properly I had to do the following behind the view otherwise i'd have 10 survived objects for MediaView as well....

    public partial class MediaView : UserControl
    {
        public MediaView()
        {
            InitializeComponent();

            MediaPlayer.MediaClosed += MediaPlayer_MediaClosed;
        }

        private async void MediaPlayer_MediaClosed(object sender, System.EventArgs e)
        {
            MediaPlayer.MediaClosed -= MediaPlayer_MediaClosed;
            MediaPlayer.ClearValue(Unosquare.FFME.MediaElement.SourceProperty);
            await MediaPlayer.Close();
            ((Grid)MediaPlayer.Parent).Children.Remove(MediaPlayer);
            MediaPlayer = null;
            GC.Collect();
        }
    }

**Comment by AaronAnz** [comment]
> > I can try to remove all the higher level disposes now that it isn't required, in order to see if that helps resolve the issue. I can try in the next week to get a minimal example.
> > I'm not able to build master currently because of `warning as errors` that block the build. I tried disabling it, I have previously had stylecop on and it seems to be the issue. I'll look at it again but I'm under a crunch this week.
> > ...Feature request (automated prerelease builds off master?)
> 
> Any updates here? I am not getting the effects of memory increases or memory leak errors in the output. Memory returns back to normal after closing -- sometimes to even lower.
> 
> ![image](https://user-images.githubusercontent.com/351978/58763428-45529b00-8520-11e9-89d6-0b7ec5ea7a8f.png)


mariodivece are you using an HLS stream? I am passing an m3u8 to the controls source property on view XAML. I'm experiencing a memory leak like they describe via the latest nuget package (4.1.300) as of this comment. I am not manually pre-loading ffmpeg at all. For sake of a test I made my app open my "MediaView" with a random twitch stream 10 times and this is what is observed. I'm using Ninject and MahApps in case those may interfere? Doubt it but figured i'd mention it.

![FFME_LEAK](https://user-images.githubusercontent.com/27394920/60068382-5b88ec80-96c3-11e9-9c87-d55baeb4e3b1.PNG)

```
   <Grid Grid.Row="0">
        <ffme:MediaElement x:Name="MediaPlayer" LoadedBehavior="Play" UnloadedBehavior="Manual" 
                            Source="{Binding Path=PlaylistSource.Source, Converter={StaticResource StringToUriConverter}}"/>       
    </Grid>
```
In order to even get my MediaView to dispose properly I had to do the following behind the view otherwise i'd have 10 survived objects for MediaView as well....

    public partial class MediaView : UserControl
    {
        public MediaView()
        {
            InitializeComponent();

            MediaPlayer.MediaClosed += MediaPlayer_MediaClosed;
        }

        private async void MediaPlayer_MediaClosed(object sender, System.EventArgs e)
        {
            MediaPlayer.MediaClosed -= MediaPlayer_MediaClosed;
            MediaPlayer.ClearValue(Unosquare.FFME.MediaElement.SourceProperty);
            await MediaPlayer.Close();
            ((Grid)MediaPlayer.Parent).Children.Remove(MediaPlayer);
            MediaPlayer = null;
            GC.Collect();
        }
    }

**Comment by mariodivece** [comment]
I don't see you are disposing of anything here and I don't see a problem with the screenshots. I also tried with twitch streams and the observed behavior is normal. The 10 surviving objects you see belong to the MediaElement's command processor which is needed to be left alive to process new commands (like open). If you would really like the CLR to return your memory immediately (not really necessary) you can just call GC.Collect after close as you are doing. Now if you see error messages in the output window of unmanaged references being left alive, then that is a memory leak.

**Comment by AaronAnz** [comment]
I see my memory go up and never release. There are 10 instances of
mediaelement being left alive. There is something wrong with your
control....

**Comment by AaronAnz** [comment]
In dotmemory after running my example my memor my is doubled and no matter
how many times you call the GC your control never release that memory. In
task manager my app can get up to 200mb just after 10 instances of media
element opening and it never goes back down no matter how many times GC is
manually called. If that's not a memory leak what is it? Also, if you truly
believe that there should be an instance left alive of mediaeleme t each
time .close is called you're insane. I'm going to have to abandon this
control of this is your desired (wrong) functionality

**Comment by AaronAnz** [comment]
What do you mean you dont see me disposing of anything here? My MediaView contains the mediaelement. There are no survived MediaViews after collection yet there are 10 MediaElements after collection. Your control is consistently increasing memory after each new stream played. 3 people are telling you this and you're acting like Visual Studio .Net Profiler can do no wrong... The "Survived Objects" column i'm highlighting is after garbage collection. My view disposes but your control wont... Blame everything but your control all you want.

**Comment by Mitra-M** [comment]
Hello All

I have the same problem and can confirm memory leak.

my  code to reproduce :

`private async void BeforePlayBtn_Click(object sender, RoutedEventArgs e)
                      {
                                 GC.Collect();
                                 GC.WaitForPendingFinalizers();
                                 GC.Collect();
                                 textblock1.Text = "Before Playing : " + GC.GetTotalMemory(true).ToString() + " bytes";
                                 await Media1.Open(new Uri(@"C:\Samples\s.avi"));
                                 await Media1.Play();
                      }
                      private  void WhilePlayingBtn_Click(object sender, RoutedEventArgs e)
                      {
                                  GC.Collect();
                                   GC.WaitForPendingFinalizers();
                                   GC.Collect();
                                   textblock2.Text = "While Playing : " + GC.GetTotalMemory(true).ToString() + " bytes";
                       }

                      private async void WhenIsclosedBtn_Click(object sender, RoutedEventArgs e)
                      {
                               if(Media1 != null)
                                 {
                                            await Media1.Stop();
                                            await Media1.Close();
                                            Media1 = null;
                                 }
                                 await Dispatcher.InvokeAsync(() =>
                                 {
                                    GC.Collect();
                                    GC.WaitForPendingFinalizers();
                                    GC.Collect();
                                    textblock3.Text = "When is closed : " + GC.GetTotalMemory(true).ToString()+" bytes";
                                 }, System.Windows.Threading.DispatcherPriority.ApplicationIdle);
                      }`



Result:
Before Playing : 6555472 bytes
While Playing : 7481496 bytes
When is closed : 7321072 bytes

**Comment by kealist** [comment]
@mariodivece I'm still on hold on that project and haven't had personal time to inspect removing the `Dispose` calls from the parent controls, so I cannot give an answer yet.  I am locked into commit e66028d567ab708dca5df954375e09fa82e4c120 for stability until I do so, so I'm far behind master.   I'll let you know when I have a chance to test, it is not likely in the immediate future.

**Comment by mariodivece** [comment]
@kealist I have added back the Dispose() method. This will get automatically called when UnloadedBehavior is set to Close. Keep in mid Dispose() will not get called if UnloadedBehavior is set to anything else.

@AaronAnz I understand you get really passionate about these discussions -- and that's sometimes a good thing. Sometimes you need to be more polite and possibly, more helpful. To find a leak in this case, you need to compare what happens after loading a file and after closing. Take memory snapshots multiple times and inspect whether the memory goes up and down on these operations and ends up at approximately the same values for each time. Keep in mind that internally, FFmpeg (a big library) gets loaded (only once) when the first file is loaded. I am sharing my results of the latest master using the sample application.

![image](https://user-images.githubusercontent.com/351978/60402500-434b1e80-9b56-11e9-9042-5f6a7ac42585.png)

As you can see, opening and closing the same file multiple times has the following ranges:
When Closed: 154.01 (first) to 148.15 (last)
When Opened: 173.16 (first) to 173.97 (last)

The above ranges do not seem to signal a memory leak. The differences in object count and the variations of a couple of MBs in memory allocations are perfectly normal -- especially for a WPF app. Remember the garbage collector is non-deterministic and variations are expected.

If you would like to drop the use of this control that is always your choice. If you want to help and provide more test results, that would be great -- but please make sure you build from latest master.

@Mitra-M Could you provide the results of more open and close operations in a row? Remember that the first time a file is loaded, ffmpeg is loaded as well -- that's probably why you don't see the memory go all the way back to the initial case. Please try with the latest master branch.

Thanks all!

**Comment by AaronAnz** [comment]
@mariodivece I literally showed you snapshots of before opening the view with your control and and after. Then I did it with the same data, same view, everything being passed to it is the same except your control is commented out. My app retained 10mb in memory after that is never released. But whatever. You dont seem to provide any better way for cleanup and then claim its all fine and dandy. Its not. I wonder why you're changing how things dispose of? Its kind of like you found an issue.

**Comment by mariodivece** [comment]
I just told you that FFmpeg needs to be statically loaded. You posted this image:
![image](https://user-images.githubusercontent.com/351978/60402883-0b92a580-9b5b-11e9-8b53-20e8b86400e5.png)

Those 10 MB of handles (8160 objects) are the delegates for the entry points of the unmanaged libraries -- i.e. FFmpeg itself -- and cannot be unloaded. The good news is that these are loaded only once and not for every instance of the MediaElement.

Support for completely unloading FFmpeg needs implementation by the FFmpeg.AutoGen project. There is already an issue documented here: #320 

It is still unclear to me why anyone would want to load FFmpeg libraries (which is expensive CPU-wise) and unload them every time media playback is needed in the application. It would take much longer for playback to start and would most likely make the UI non-responsive in short bursts.

Adding the ability to unload FFmpeg libraries is certainly not a high priority in this project.

As for the dispose changes, these are related  only to the UnloadedBehavior property. It won't unload the unmanaged libraries.

**Comment by AaronAnz** [comment]
Well I had 10 lingering references to mediaelement no matter what I tried.
So idk what you want from me?

**Comment by mariodivece** [comment]
You are correct on the lingering references. UnloadedBehavior is broken in the current nuget. 
This should be fixed in my latest commit (the one you commented about).

**Comment by mariodivece** [comment]
Closing this issue as this package is getting released. Please file a new issue if you encounter further problems. Dispose is back and UnloadedBehavior is now working as expected according to my tests.

**Comment by AaronAnz** [comment]
I was running it from the latest nuget. You claimed there was no issue and now you're saying there was an issue. As well, once you updated nuget, my 15,000 survived objects for your lib dropped to 138. I don't believe it was "just the ffmpeg bindings". We can agree to disagree.

**Comment by manity42** [comment]
Hello when should this dispose feature best be used in WPF, would it be used when changing from 1 file to another in the element or should it only be used on application exit in Dispatcher.shutdown event

**Comment by mariodivece** [comment]
Don't use it at all if the MediaElement control is always on the visual tree. 
Don't call dispose of your MediaElement if the Unloaded behavior is set to Close (the default). If you want to call dispose manually please set the Unloaded Behavior to Manual.

**Comment by Mitra-M** [comment]
@mariodivece, You are right.

I learned something new (new for me) about garbage. Thanks a lot.


for both standard WPF and FFME I made 2 following projects:

WpfApp1:

       `<Window x:Class="WpfApp1.MainWindow"
             xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             xmlns:ffme="clr-namespace:Unosquare.FFME;assembly=ffme.win"
             mc:Ignorable="d"
             Title="FFME Memory Test" Height="450" Width="800">
           <Grid>
                      <Grid.RowDefinitions>
                                 <RowDefinition Height="200"/>
                                 <RowDefinition  Height="25"/>
                                 <RowDefinition  />
                      </Grid.RowDefinitions>
                      <Grid.ColumnDefinitions>
                                 <ColumnDefinition/>
                                 <ColumnDefinition Width="10"/>
                                 <ColumnDefinition/>
                      </Grid.ColumnDefinitions>
                      <ffme:MediaElement x:Name="Media1" Background="Silver"  LoadedBehavior="Manual" UnloadedBehavior="Manual" Grid.ColumnSpan="3"/>
                      <Button Content="Open" Click="PlayBtn_Click"  Grid.Row="1" Grid.Column="0"   />
                      <Button Content="Close" Click="ClosedBtn_Click"  Grid.Row="1"   Grid.Column="2"   />
                      <TextBox Name="text1"  Grid.Row="2"   />
                      <TextBox x:Name="text2"  Grid.Row="2"  Grid.Column="2" />
           </Grid>
     </Window>
    c# code :
     using System;
     using System.Windows;
     using System.Windows.Input;
     using System.Windows.Threading;
     namespace WpfApp1
    {
           public partial class MainWindow :Window
           {
                      public MainWindow()
                      {
                                 InitializeComponent();
                      }
                      int OpenCounter,CloseCounter;
                      private async void PlayBtn_Click(object sender, RoutedEventArgs e)
                      {
                                 Mouse.OverrideCursor = Cursors.AppStarting;
                                 await Media1.Open(new Uri(@"C:\Samples\s.avi"));
                                 await Media1.Play();

                                 DispatcherTimer t = new DispatcherTimer { Interval = TimeSpan.FromSeconds(2) };
                                 EventHandler h = null;
                                 h = (delegate
                                 {
                                            t.Stop(); t.Tick -= h; t = null;h = null;
                                            GC.Collect();
                                            GC.WaitForPendingFinalizers();
                                            GC.Collect();
                                            text1.Text += "Opened" + (OpenCounter++) + " :" + (GC.GetTotalMemory(true) / 1024).ToString() + " KB\n";
                                            Mouse.OverrideCursor = null;
                                 });
                                 t.Tick += h;
                                 t.Start();
                      }

                      private async void ClosedBtn_Click(object sender, RoutedEventArgs e)
                      {
                                 Mouse.OverrideCursor = Cursors.AppStarting;
                                 if(Media1 != null)
                                 {
                                            await Media1.Stop();
                                            await Media1.Close();
                                            Media1.Source = null;
                                 }
                                 DispatcherTimer t = new DispatcherTimer { Interval = TimeSpan.FromSeconds(2) };
                                 EventHandler h = null;
                                 h = (delegate
                                 {
                                            t.Stop(); t.Tick -= h; t = null; h = null;
                                            GC.Collect();
                                            GC.WaitForPendingFinalizers();
                                            GC.Collect();
                                            text2.Text += "Closed" + (CloseCounter++) + " :" + (GC.GetTotalMemory(true) / 1024).ToString() + " KB\n";
                                            Mouse.OverrideCursor = null;
                                 });
                                 t.Tick += h;
                                 t.Start();
                      }
               }
      }
   `

WpfApp2:

     `      <Window x:Class="WpfApp2.MainWindow"
            xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d"
             Title="Standard WPF Memory Test" Height="450" Width="800">
           <Grid>
                      <Grid.RowDefinitions>
                                 <RowDefinition Height="200"/>
                                 <RowDefinition  Height="25"/>
                                 <RowDefinition  />
                      </Grid.RowDefinitions>
                      <Grid.ColumnDefinitions>
                                 <ColumnDefinition/>
                                 <ColumnDefinition Width="10"/>
                                 <ColumnDefinition/>
                      </Grid.ColumnDefinitions>
                      <MediaElement x:Name="Media1" LoadedBehavior="Manual" UnloadedBehavior="Manual" Grid.ColumnSpan="3"/>
                      <Button Content="Open" Click="PlayBtn_Click"  Grid.Row="1" Grid.Column="0"   />
                      <Button Content="Close" Click="ClosedBtn_Click"  Grid.Row="1"   Grid.Column="2"   />
                      <TextBox Name="text1"  Grid.Row="2"   />
                      <TextBox x:Name="text2"  Grid.Row="2"  Grid.Column="2" />
           </Grid>
     </Window>

       c# code : 

      using System;
      using System.Windows;
      using System.Windows.Input;
      using System.Windows.Threading;
      namespace WpfApp2
      {
           public partial class MainWindow :Window
           {
                      public MainWindow()
                      {
                                 InitializeComponent();
                      }

                      int OpenCounter, CloseCounter;
                      private void PlayBtn_Click(object sender, RoutedEventArgs e)
                      {
                                 Mouse.OverrideCursor = Cursors.AppStarting;
                                 Media1.Source=new Uri(@"C:\Samples\s.avi");
                                 Media1.Play();

                                 DispatcherTimer t = new DispatcherTimer { Interval = TimeSpan.FromSeconds(2) };
                                 EventHandler h = null;
                                 h = (delegate
                                 {
                                            t.Stop(); t.Tick -= h; t = null; h = null;
                                            GC.Collect();
                                            GC.WaitForPendingFinalizers();
                                            GC.Collect();
                                            text1.Text += "Opened" + (OpenCounter++) + " :" + (GC.GetTotalMemory(true) / 1024).ToString() + " KB\n";
                                            Mouse.OverrideCursor = null;
                                 });
                                 t.Tick += h;
                                 t.Start();
                      }

                      private void ClosedBtn_Click(object sender, RoutedEventArgs e)
                      {
                                 Mouse.OverrideCursor = Cursors.AppStarting;
                                 if(Media1 != null)
                                 {
                                             Media1.Stop();
                                             Media1.Close();
                                            Media1.Source = null;
                                 }
                                 DispatcherTimer t = new DispatcherTimer { Interval = TimeSpan.FromSeconds(2) };
                                 EventHandler h = null;
                                 h = (delegate
                                 {
                                            t.Stop(); t.Tick -= h; t = null; h = null;
                                            GC.Collect();
                                            GC.WaitForPendingFinalizers();
                                            GC.Collect();
                                            text2.Text += "Closed" + (CloseCounter++) + " :" + (GC.GetTotalMemory(true) / 1024).ToString() + " KB\n";
                                            Mouse.OverrideCursor = null;
                                 });
                                 t.Tick += h;
                                 t.Start();
                      }
           }
     }
    `

The results:
              FFMediaelement                                             WPF Mediaelement
Opened0 :3847 KB   Closed0 :3713 KB           Opened0 :1611 KB   Closed0 :1620 KB
Opened1 :3878 KB   Closed1 :3723 KB           Opened1 :1622 KB   Closed1 :1624 KB
Opened2 :3883 KB   Closed2 :3724 KB           Opened2 :1626 KB   Closed2 :1628 KB
Opened3 :3889 KB   Closed3 :3728 KB           Opened3 :1630 KB   Closed3 :1634 KB
Opened4 :3896 KB   Closed4 :3734 KB           Opened4 :1636 KB   Closed4 :1637 KB
Opened5 :3890 KB   Closed5 :3737 KB           Opened5 :1641 KB   Closed5 :1643 KB


Both behave the same way, and this shows that FFme is working properly.

Thanks again for this great project.
